### ISSUE TITLE
Broken text to floating point conversion on some locales

### ISSUE THREAD

**Issue by JulianGro** [jkuhlmann/cgltf#259]
The CGLTF_ATOF function seems to only work properly on systems with `.` as decimal separator. On German Linux systems for example `,` is used as decimal separator, which breaks the conversion. My understanding is that this is a problem with the C standard being decimal separator dependent in the first place.

In our C++ code, we work around this the following way:
```cpp
float atof_locale_independent(char* str);

#define CGLTF_ATOF(str) atof_locale_independent(str)
#include <cgltf.h>

float atof_locale_independent(char* str) {
    //TODO: Once we have C++17 we can use std::from_chars
    std::istringstream streamToParse(str);
    streamToParse.imbue(std::locale("C"));
    float value;
    if (!(streamToParse >> value)) {
        qDebug(modelformat) << "cgltf: Cannot parse float from string: " << str;
        return 0.0f;
    }
    return value;
}
```

Here is an example model which completely breaks on systems with `,` decimal separator: http://oaktown.pl/models/stopnie_2_1.gltf (model is CC0)

**Comment by oreo639** [comment]
Btw, to reproduce this you need to do `setlocale(LC_ALL, "");` first (since POSIX declares that the locale always defaults to `C`).

After that printing the atof calls confirms that the decimals are truncated.

> My understanding is that this is a problem with the C standard being decimal separator dependent in the first place.

To clarify, it is is an issue with the POSIX standard as the C standard has no concept of locales.

GNU does have `atof_l`/`strod_l` (which is not POSIX standard):
```
#define CGLTF_ATOF(str) strtod_l(str, NULL, newlocale(LC_ALL, "C", NULL))
```
It isn't really portable though since it requires `gnuXX/gnu++XX` (which is equivalent to `_GNU_SOURCE` defined before everything else with glibc), it is supported pretty widely but not everywhere (and Windows would use `_create_locale(LC_ALL,"C");` instead of `newlocale(LC_ALL, "C", NULL)`).

Also, keep in mind that calling newlocale does allocate memory, although with glibc `newlocale()` will return the preallocated C locale if base is null or LC_ALL_MASK is used: https://github.com/bminor/glibc/blob/bb6496b96444dfd55d7105396780f6eba14b1cd9/locale/newlocale.c#L75

Afaict the only portable way to handle this is to re-implement a locale-independent strtod/atof.

**Comment by oreo639** [comment]
Edit: I said that locale was a part of POSIX and not C, I was mistaken, sorry.

**Comment by JCash** [comment]
I think the cgltf.h should come with its own number parsing mechanic by default.
E.g. as described, the `atof` uses the locale which is clearly against the [json specification](https://datatracker.ietf.org/doc/html/rfc8259#section-6) which must use the `.` character.

I'm not sure which implementation to use, but something small, robust and well used should of course be preferable.

@jkuhlmann As this issue is a fairly serious issue on Linux machines, perhaps we could bump the priority on this one?
If you don't have the time, would you consider a PR for adding json number parsing?

**Comment by jkuhlmann** [comment]
@JCash Certainly, I'd consider a PR for this. Do you have one with a compatible license, or a tested custom one?
